pkg/tsdb/prometheus/prometheus.go

```diff
@@ -149,44 +129,46 @@ func (s *Service) QueryData(ctx context.Context, req *backend.QueryDataRequest)
 	}
 
 	for _, query := range queries {
-		timeRange := apiv1.Range{
-			Start: query.Start,
-			End:   query.End,
-			Step:  query.Step,
-		}
-
-		plog.Debug("Sending query", "start", timeRange.Start, "end", timeRange.End, "step", timeRange.Step, "query", query.Expr)
+		plog.Debug("Sending query", "start", query.Start, "end", query.End, "step", query.Step, "query", query.Expr)
 
 		span, ctx := opentracing.StartSpanFromContext(ctx, "datasource.prometheus")
 		span.SetTag("expr", query.Expr)
 		span.SetTag("start_unixnano", query.Start.UnixNano())
 		span.SetTag("stop_unixnano", query.End.UnixNano())
 		defer span.Finish()
 
-		var response model.Value
+		response := make(map[PrometheusQueryType]model.Value)
 
-		switch query.QueryType {
-		case Range:
-			response, _, err = client.QueryRange(ctx, query.Expr, timeRange)
+		if query.RangeQuery {
+			timeRange := apiv1.Range{
+				Step: query.Step,
+				// Align query range to step. It rounds start and end down to a multiple of step.
+				Start: time.Unix(int64(math.Floor((float64(query.Start.Unix()+query.UtcOffsetSec)/query.Step.Seconds()))*query.Step.Seconds()-float64(query.UtcOffsetSec)), 0),
```

Comment:
I would probably mention here that we do not align for things like day light saving etc because I assume prometheus also does not do that? https://prometheus.io/docs/prometheus/latest/querying/basics/#time-durations

Reply:
👍 sounds good, I'll update

Reply:
Hmmm, thinking about it, I am not sure why we are not doing it. I would have to look into it closer and I don't want to add comment which might be falsy/inaccurate. I have created an [item](https://github.com/grafana/grafana/issues/37784) in epic to get back to this. This is how we do it on frontend, so there is no regression.